lib/mssql.js

```diff
@@ -0,0 +1,215 @@
+      })
+    ),
+};
+
+export function sanitizeForCellTag(string) {
+  let idx = 0;
+  const sanitized = string.split("").reduce((acc, char) => {
+    if (char === "?") {
+      idx += 1;
+      return acc.concat(`@${idx}`);
+    }
+    return acc.concat(char);
+  }, []);
+
+  return sanitized.join("");
+}
+
+export async function query(req, res, pool) {
+  const db = await pool;
+  const body = await json(req);
+
+  if (!validate(body)) throw badRequest();
+
+  res.setHeader("Content-Type", "text/plain");
+  const keepAlive = setInterval(() => res.write("\n"), 25e3);
+
+  let {sql, params = []} = body;
+  sql = sanitizeForCellTag(sql);
+
+  try {
+    await new Promise((resolve, reject) => {
+      const request = new mssql.Request(db);
+      const stream = request.toReadableStream();
+
+      params.forEach((param, idx) => {
+        request.input(`${idx + 1}`, param);
+      });
+
+      request.query(sql);
+      request.once("recordset", () => clearInterval(keepAlive));
+      request.on("recordset", (columns) => {
+        const schema = {
+          type: "array",
+          items: {
+            type: "object",
+            properties: Object.entries(columns).reduce(
+              (schema, [name, props]) => {
+                return {
+                  ...schema,
+                  ...{[name]: dataTypeSchema({type: props.type.name})},
+                };
+              },
+              {}
+            ),
+          },
+        };
+
+        res.write(`${JSON.stringify(schema)}`);
+        res.write("\n");
+      });
+
+      stream.pipe(JSONStream.stringify("", "\n", "\n")).pipe(res);
+      stream.on("done", () => {
+        resolve();
+      });
+      stream.on("error", (error) => {
+        if (!request.canceled) {
+          request.cancel();
+        }
+        reject(error);
+      });
+    });
+  } catch (error) {
+    if (!error.statusCode) error.statusCode = 400;
+    throw error;
+  } finally {
+    clearInterval(keepAlive);
+  }
+
+  res.end();
+}
+
+export async function check(req, res, pool) {
+  const db = await pool;
+
+  // See: https://learn.microsoft.com/en-us/sql/relational-databases/system-functions/sys-fn-my-permissions-transact-sql
+  const rows = await db.request().query(
+    `USE ${db.config.database};
+     SELECT * FROM fn_my_permissions (NULL, 'DATABASE');`
+  );
+
+  const grants = rows.recordset.map((rs) => rs.permission_name);
+  const permissive = grants.filter((g) => !READ_ONLY.has(g));
+
+  if (permissive.length)
+    throw failedCheck(
+      `User has too permissive grants: ${permissive.join(", ")}`
+    );
+
+  return {ok: true};
+}
+
+export default (credentials) => async (req, res) => {
+  const pool = mssqlPool.get(credentials, credentials);
+  if (req.method === "POST") {
+    if (req.url === "/check") {
+      return check(req, res, pool);
+    }
+
+    if (["/query-stream"].includes(req.url)) {
+      return query(req, res, pool);
+    }
+
+    throw notFound();
```

Comment:
Probably should be a not-implemented error. That would allow the client code to catch this and react to it differently.